### PR TITLE
Save workload results into files

### PR DESCRIPTION
**Pull request by fbalak**
Solves:

- workload fixtures can use function scope fixtures (like factory fixtures)
- lesser consumption of memory as all alerts related to tests are now loaded only when needed (it would become a problem with a lot of tests)
- easier debugging as tests can be replayed without waiting again for collection of all data
- fixed severity level for CephMgrIsAbsent alert

### CODE REVIEW

**Comment by fbalak** [comment]
Logs from test that had results loaded from file:
```
$ run-ci -svvvrx tests/manage/monitoring/prometheus/test_deployment_status.py -k test_ceph_manager_stopped --cluster-path /home/ocsqe/data/ocs-instances/current
Collecting Cluster versions
======================================================== test session starts ========================================================
platform linux -- Python 3.6.6, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/ocsqe/data/fbalak/ocs-ci/venv3/bin/python3
cachedir: .pytest_cache
metadata: {'Test Run Name': 'OCS_US_', 'Cluster Version': '4.2.0-0.nightly-2019-08-31-102058', 'Ceph Version': '14.2.2 (4f8fa0a0024755aae7d95567c63f11d6862d55be) nautilus (stable)', 'Rook Version': 'v1.0.0-597.g5c59a1e', 'csi-provisioner': 'v1.3.0', 'cephfsplugin': None, 'rbdplugin': None}
rootdir: /home/ocsqe/data/fbalak/ocs-ci, inifile: pytest.ini
plugins: reportportal-1.0.5, metadata-1.8.0, html-1.21.1, logger-0.5.0, marker-bugzilla-0.9.1.dev2
collected 2 items / 1 deselected / 1 selected                                                                                       

tests/manage/monitoring/prometheus/test_deployment_status.py::test_ceph_manager_stopped 
---------------------------------------------------------- live log setup -----------------------------------------------------------
18:03:40 - MainThread - tests.conftest - INFO - All logs located at /tmp/ocs-ci-logs-1567612986
18:03:40 - MainThread - ocs_ci.deployment.factory - INFO - Deployment key = aws_ipi
18:03:40 - MainThread - ocs_ci.deployment.factory - INFO - Current deployment platform: AWS,deployment type: ipi
18:03:40 - MainThread - ocs_ci.utility.utils - INFO - Executing command: ./bin/oc version
18:03:42 - MainThread - ocs_ci.utility.utils - INFO - OpenShift Client version: Client Version: version.Info{Major:"", Minor:"", GitVersion:"openshift-clients-4.2.0-201908281419", GitCommit:"372c08a6a95e4ecc7c61dd16052fc12f6bca376c", GitTreeState:"clean", BuildDate:"2019-08-28T19:10:30Z", GoVersion:"go1.12.8", Compiler:"gc", Platform:"linux/amd64"}
                                                      Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.0+a91f2ac", GitCommit:"a91f2ac", GitTreeState:"clean", BuildDate:"2019-08-28T23:03:31Z", GoVersion:"go1.12.8", Compiler:"gc", Platform:"linux/amd64"}
                                                      OpenShift Version: 4.2.0-0.nightly-2019-08-31-102058
18:03:42 - MainThread - ocs_ci.utility.utils - INFO - The rook directory /home/ocsqe/data/fbalak/ocs-ci/data/rook already exists, ocs-ci will skip the clone of rook repository.
18:03:42 - MainThread - ocs_ci.utility.utils - INFO - Fetching latest changes from rook repository.
18:03:42 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git fetch --all
18:03:44 - MainThread - ocs_ci.utility.utils - INFO - Checkout rook repository to specific branch: master
18:03:44 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git checkout master
18:03:44 - MainThread - ocs_ci.utility.utils - WARNING - Command warning: Already on 'master'
18:03:44 - MainThread - ocs_ci.utility.utils - INFO - Reset branch: master with latet changes
18:03:44 - MainThread - ocs_ci.utility.utils - INFO - Executing command: git reset --hard origin/master
18:03:44 - MainThread - tests.manage.monitoring.conftest - INFO - Using measurement dir from configuration: /tmp/pytest-of-ocsqe/pytest-12/measurement_results
18:03:44 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ocsqe/data/ocs-instances/current/auth/kubeconfig get Deployment --selector=app=rook-ceph-mgr -o yaml
18:03:47 - MainThread - tests.manage.monitoring.conftest - INFO - File /tmp/pytest-of-ocsqe/pytest-12/measurement_results/stop_mgr.json already created. Trying to use it for tests...
18:03:47 - MainThread - tests.manage.monitoring.conftest - INFO - File /tmp/pytest-of-ocsqe/pytest-12/measurement_results/stop_mgr.json loaded. Content of file:
                                                                  {'start': 1567612522.075062, 'stop': 1567612887.2581115, 'result': {'apiVersion': 'extensions/v1beta1', 'kind': 'Deployment', 'metadata': {'annotations': {'deployment.kubernetes.io/revision': '1'}, 'creationTimestamp': '2019-09-02T08:29:06Z', 'generation': 10, 'labels': {'app': 'rook-ceph-mgr', 'ceph-version': '14.2.2', 'ceph_daemon_id': 'a', 'instance': 'a', 'mgr': 'a', 'rook-version': 'v1.0.0-597.g5c59a1e', 'rook_cluster': 'openshift-storage'}, 'name': 'rook-ceph-mgr-a', 'namespace': 'openshift-storage', 'ownerReferences': [{'apiVersion': 'ceph.rook.io/v1', 'blockOwnerDeletion': True, 'kind': 'CephCluster', 'name': 'rook-ceph', 'uid': '2faae07a-cd5b-11e9-97a3-06ceef04ea0a'}], 'resourceVersion': '1514978', 'selfLink': '/apis/extensions/v1beta1/namespaces/openshift-storage/deployments/rook-ceph-mgr-a', 'uid': 'ba3c60ba-cd5b-11e9-b07b-02abd5a97892'}, 'spec': {'progressDeadlineSeconds': 600, 'replicas': 0, 'revisionHistoryLimit': 10, 'selector': {'matchLabels': {'app': 'rook-ceph-mgr', 'ceph_daemon_id': 'a', 'instance': 'a', 'mgr': 'a', 'rook_cluster': 'openshift-storage'}}, 'strategy': {'type': 'Recreate'}, 'template': {'metadata': {'annotations': {'prometheus.io/port': '9283', 'prometheus.io/scrape': 'true'}, 'creationTimestamp': None, 'labels': {'app': 'rook-ceph-mgr', 'ceph_daemon_id': 'a', 'instance': 'a', 'mgr': 'a', 'rook_cluster': 'openshift-storage'}, 'name': 'rook-ceph-mgr-a'}, 'spec': {'affinity': {}, 'containers': [{'args': ['--fsid=1e937049-5b98-4166-a3fb-28e0b8b1e3c2', '--keyring=/etc/ceph/keyring-store/keyring', '--log-to-stderr=true', '--err-to-stderr=true', '--mon-cluster-log-to-stderr=true', '--log-stderr-prefix=debug ', '--default-log-to-file=false', '--default-mon-cluster-log-to-file=false', '--mon-host=$(ROOK_CEPH_MON_HOST)', '--mon-initial-members=$(ROOK_CEPH_MON_INITIAL_MEMBERS)', '--id=a', '--setuser=ceph', '--setgroup=ceph', '--client-mount-uid=0', '--client-mount-gid=0', '--foreground', '--public-addr=$(ROOK_POD_IP)'], 'command': ['ceph-mgr'], 'env': [{'name': 'CONTAINER_IMAGE', 'value': 'ceph/ceph:v14.2.2-20190722'}, {'name': 'POD_NAME', 'valueFrom': {'fieldRef': {'apiVersion': 'v1', 'fieldPath': 'metadata.name'}}}, {'name': 'POD_NAMESPACE', 'valueFrom': {'fieldRef': {'apiVersion': 'v1', 'fieldPath': 'metadata.namespace'}}}, {'name': 'NODE_NAME', 'valueFrom': {'fieldRef': {'apiVersion': 'v1', 'fieldPath': 'spec.nodeName'}}}, {'name': 'POD_MEMORY_LIMIT', 'valueFrom': {'resourceFieldRef': {'divisor': '0', 'resource': 'limits.memory'}}}, {'name': 'POD_MEMORY_REQUEST', 'valueFrom': {'resourceFieldRef': {'divisor': '0', 'resource': 'requests.memory'}}}, {'name': 'POD_CPU_LIMIT', 'valueFrom': {'resourceFieldRef': {'divisor': '1', 'resource': 'limits.cpu'}}}, {'name': 'POD_CPU_REQUEST', 'valueFrom': {'resourceFieldRef': {'divisor': '0', 'resource': 'requests.cpu'}}}, {'name': 'ROOK_CEPH_MON_HOST', 'valueFrom': {'secretKeyRef': {'key': 'mon_host', 'name': 'rook-ceph-config'}}}, {'name': 'ROOK_CEPH_MON_INITIAL_MEMBERS', 'valueFrom': {'secretKeyRef': {'key': 'mon_initial_members', 'name': 'rook-ceph-config'}}}, {'name': 'ROOK_OPERATOR_NAMESPACE', 'value': 'openshift-storage'}, {'name': 'ROOK_CEPH_CLUSTER_CRD_VERSION', 'value': 'v1'}, {'name': 'ROOK_VERSION', 'value': 'v1.0.0-597.g5c59a1e'}, {'name': 'ROOK_CEPH_CLUSTER_CRD_NAME', 'value': 'openshift-storage'}, {'name': 'ROOK_POD_IP', 'valueFrom': {'fieldRef': {'apiVersion': 'v1', 'fieldPath': 'status.podIP'}}}], 'image': 'ceph/ceph:v14.2.2-20190722', 'imagePullPolicy': 'IfNotPresent', 'lifecycle': {'postStart': {'exec': {'command': ['chown', '--recursive', '--verbose', 'ceph:ceph', '/var/log/ceph']}}}, 'livenessProbe': {'failureThreshold': 3, 'httpGet': {'path': '/', 'port': 9283, 'scheme': 'HTTP'}, 'initialDelaySeconds': 60, 'periodSeconds': 10, 'successThreshold': 1, 'timeoutSeconds': 1}, 'name': 'mgr', 'ports': [{'containerPort': 6800, 'name': 'mgr', 'protocol': 'TCP'}, {'containerPort': 9283, 'name': 'http-metrics', 'protocol': 'TCP'}, {'containerPort': 8443, 'name': 'dashboard', 'protocol': 'TCP'}], 'resources': {}, 'securityContext': {'privileged': True}, 'terminationMessagePath': '/dev/termination-log', 'terminationMessagePolicy': 'File', 'volumeMounts': [{'mountPath': '/etc/ceph', 'name': 'rook-config-override', 'readOnly': True}, {'mountPath': '/etc/ceph/keyring-store/', 'name': 'rook-ceph-mgr-a-keyring', 'readOnly': True}, {'mountPath': '/var/log/ceph', 'name': 'rook-ceph-log'}, {'mountPath': '/var/lib/ceph/mgr/ceph-a', 'name': 'ceph-daemon-data'}]}], 'dnsPolicy': 'ClusterFirst', 'restartPolicy': 'Always', 'schedulerName': 'default-scheduler', 'securityContext': {}, 'serviceAccount': 'rook-ceph-mgr', 'serviceAccountName': 'rook-ceph-mgr', 'terminationGracePeriodSeconds': 30, 'volumes': [{'configMap': {'defaultMode': 420, 'items': [{'key': 'config', 'mode': 292, 'path': 'ceph.conf'}], 'name': 'rook-config-override'}, 'name': 'rook-config-override'}, {'name': 'rook-ceph-mgr-a-keyring', 'secret': {'defaultMode': 420, 'secretName': 'rook-ceph-mgr-a-keyring'}}, {'hostPath': {'path': '/var/lib/rook/openshift-storage/log', 'type': ''}, 'name': 'rook-ceph-log'}, {'emptyDir': {}, 'name': 'ceph-daemon-data'}]}}}, 'status': {'conditions': [{'lastTransitionTime': '2019-09-02T08:29:06Z', 'lastUpdateTime': '2019-09-02T08:29:09Z', 'message': 'ReplicaSet "rook-ceph-mgr-a-697486687d" has successfully progressed.', 'reason': 'NewReplicaSetAvailable', 'status': 'True', 'type': 'Progressing'}, {'lastTransitionTime': '2019-09-04T15:52:59Z', 'lastUpdateTime': '2019-09-04T15:52:59Z', 'message': 'Deployment has minimum availability.', 'reason': 'MinimumReplicasAvailable', 'status': 'True', 'type': 'Available'}], 'observedGeneration': 10}}, 'metadata': None, 'prometheus_alerts': [{'labels': {'alertname': 'Watchdog', 'severity': 'none'}, 'annotations': {'message': 'This is an alert meant to ensure that the entire alerting pipeline is functional.\nThis alert is always firing, therefore it should always be firing in Alertmanager\nand always fire against a receiver. There are integrations with various notification\nmechanisms that send a notification when this alert is not firing. For example the\n"DeadMansSnitch" integration in PagerDuty.\n'}, 'state': 'firing', 'activeAt': '2019-09-02T08:17:30.163677339Z', 'value': '1e+00'}, {'labels': {'alertname': 'TargetDown', 'job': 'rook-ceph-mgr', 'severity': 'warning'}, 'annotations': {'message': '100% of the rook-ceph-mgr targets are down.'}, 'state': 'pending', 'activeAt': '2019-09-04T15:55:00.163677339Z', 'value': '1e+02'}, {'labels': {'alertname': 'CephMgrIsAbsent', 'severity': 'critical'}, 'annotations': {'description': 'Ceph Manager has disappeared from Prometheus target discovery.', 'message': 'Storage metrics collector service not available anymore.', 'severity_level': 'critical', 'storage_type': 'ceph'}, 'state': 'pending', 'activeAt': '2019-09-04T15:55:22.047206658Z', 'value': '1e+00'}, {'labels': {'alertname': 'CephMgrIsAbsent', 'severity': 'critical'}, 'annotations': {'description': 'Ceph Manager has disappeared from Prometheus target discovery.', 'message': 'Storage metrics collector service not available anymore.', 'severity_level': 'critical', 'storage_type': 'ceph'}, 'state': 'firing', 'activeAt': '2019-09-04T15:55:22.047206658Z', 'value': '1e+00'}]}
18:03:47 - MainThread - tests.manage.monitoring.conftest - INFO - Upscaling deployment rook-ceph-mgr-a back to 1
18:03:47 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ocsqe/data/ocs-instances/current/auth/kubeconfig scale --replicas=1 deployment/rook-ceph-mgr-a
----------------------------------------------------------- live log call -----------------------------------------------------------
18:03:49 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc login -u kubeadmin -p pDrkk-4iQYN-DZTYE-D6eiW
18:03:54 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-monitoring --kubeconfig /home/ocsqe/data/ocs-instances/current/auth/kubeconfig whoami --show-token
18:03:55 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-monitoring --kubeconfig /home/ocsqe/data/ocs-instances/current/auth/kubeconfig get Route prometheus-k8s -o yaml
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - Generated CA certification file: /tmp/tmpa8x7apwd
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - Checking properties of found CephMgrIsAbsent alerts
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - Alerts were triggered correctly during utilization
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - Checking for CephMgrIsAbsent alerts. There should be no alerts ... 0 found
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - Cleared alerts: []
18:03:58 - MainThread - /home/ocsqe/data/fbalak/ocs-ci/ocs_ci/utility/prometheus.py - INFO - CephMgrIsAbsent alerts were cleared
PASSED

============================================== 1 passed, 1 deselected in 18.79 seconds ==============================================
```

File: tests/manage/monitoring/conftest.py [hunk 1]
```diff
@@ -64,51 +73,98 @@ def prometheus_log(info, alert_list):
...
+    return results
+
+
+@pytest.fixture
+def measurement_dir(tmp_path):
+    """
+    Returns directory path where should be stored all results related
+    to measurement. If 'measurement_dir' is provided by config then use it,
+    otherwise new directory is generated.
+
```
Comment (ebenahar) [inline tests/manage/monitoring/conftest.py]: Please add Args

File: tests/manage/monitoring/conftest.py [hunk 2]
```diff
@@ -139,14 +195,15 @@ def stop_mgr():
...
+    test_file = os.path.join(measurement_dir, 'stop_mgr.json')
+    measured_op = measure_operation(stop_mgr, test_file)
     logger.info(f"Upscaling deployment {mgr} back to 1")
     oc.exec_oc_cmd(f"scale --replicas=1 deployment/{mgr}")
     return measured_op
 
 
-@pytest.fixture(scope="session")
-def workload_stop_ceph_mon():
+@pytest.fixture
```
Comment (ebenahar) [inline tests/manage/monitoring/conftest.py]: how is @pytest.fixture different than @pytest.fixture() in terms of the scope?

File: tests/manage/monitoring/conftest.py [hunk 3]
```diff
@@ -64,51 +73,98 @@ def prometheus_log(info, alert_list):
...
-
-
-@pytest.fixture(scope="session")
-def workload_stop_ceph_mgr():
+        # init logging thread that checks for Prometheus alerts
+        # while workload is running
+        # based on https://docs.python.org/3/howto/logging-cookbook.html#logging-from-multiple-threads
+        info = {'run': True}
+        alert_list = []
+
```
Comment (ebenahar) [inline tests/manage/monitoring/conftest.py]: Please add comments in the code in this function

Comment (fbalak) [reply tests/manage/monitoring/conftest.py]: I added some comments. Please let me know if there needs to be added more comments to more parts of code.

File: tests/manage/monitoring/conftest.py [hunk 4]
```diff
@@ -139,14 +199,15 @@ def stop_mgr():
         time.sleep(run_time)
         return oc.get(mgr)
 
-    measured_op = measure_operation(stop_mgr)
+    test_file = os.path.join(measurement_dir, 'stop_mgr.json')
```
Comment (mbukatov) [inline tests/manage/monitoring/conftest.py]: To improve clarify, could you use full name of the fixture for the filename? Eg. `workload_stop_ceph_mgr.json`.

File: tests/manage/monitoring/conftest.py [hunk 5]
```diff
@@ -196,7 +257,8 @@ def stop_mon():
         time.sleep(run_time)
         return mons_to_stop
 
-    measured_op = measure_operation(stop_mon)
+    test_file = os.path.join(measurement_dir, 'stop_mon.json')
```
Comment (mbukatov) [inline tests/manage/monitoring/conftest.py]: The same suggestion about the filename.

File: tests/manage/monitoring/conftest.py [hunk 6]
```diff
@@ -64,51 +73,102 @@ def prometheus_log(info, alert_list):
...
+        logger.info(
+            f"Using measurement dir from configuration: {measurement_dir}"
+        )
+    else:
+        measurement_dir = os.path.join(
+            os.path.dirname(tmp_path),
+            'measurement_results'
+        )
+        if not os.path.exists(measurement_dir):
+            os.mkdir(measurement_dir)
```
Comment (mbukatov) [inline tests/manage/monitoring/conftest.py]: Could this actually happen? Wouldn't it be more useful in the 1st branch of this if block (eg. when one specifies the measurement dir, but it doesn't exist)?

File: tests/manage/monitoring/conftest.py (hunk 4 above)
Comment (fbalak) [reply tests/manage/monitoring/conftest.py]: done

File: tests/manage/monitoring/conftest.py (hunk 5 above)
Comment (fbalak) [reply tests/manage/monitoring/conftest.py]: done

File: tests/manage/monitoring/conftest.py (hunk 6 above)
Comment (fbalak) [reply tests/manage/monitoring/conftest.py]: That makes sense. I made the part with creating directory if it doesn't exist accessible to both branches.

**Pull request merged**